Fixing bug in spin treatment for PdosPlot

pfebrer opened the pull request:
There were two bugs:
- `pdos_plot.split_DOS(on="spin")` didn't work properly. The same bug caused `pdos_plot.split_DOS(...)` to cancel the ability to use the spin parameter in requests anymore.
- PDOS calculated from the hamiltonian was always interpreted to be spin polarized.

The first bug was not detected by the tests because there is no test that involves spin. Probably I should generate the `SrTiO3.PDOS` file from a spin polarized calculation. I will do it tomorrow, and then I will also need to change something in the `test_pdos.py` file.

The second bug sneaked through tests due to a magical compensation by the first bug.

Also, as @arsalan-akhtar reported this bug, he also noted (I already had it in mind for a long time) that it is usual to have "spin down DOS" with a negative value. I introduced a `scale` parameter to the request dict so that the user now can multiply the DOS by whatever number they want:

```python
pdos_plot.update_settings(requests=[
    {"spin": [0]},
    {"spin": [1], "scale": -1}
])
```

in the future there should probably be an "invert_spin_down" setting that automatically sets up the scale values, but for now this should be fine.

**PS**: I'm in the process of thinking about the things of your PR, but this bug was quite important and it took me a while to realize what was going on so I wanted to fix it while it was fresh in my mind.

zerothi commented:
This is perfect.
I much prefer smaller PR's that fixes individual problems, so definitely the right way of doing things!

I would suggest you put the scale also for the first spin channel. It may be useful for other things as well. So why not have it both places? :)

zerothi commented on sisl/viz/plotly/plots/pdos.py:
```diff
@@ -177,6 +183,13 @@ class PdosPlot(Plot):
+                    key="scale", name="Scale",
+                    default=1,
+                    params={"min": None},
+                    help="The final DOS will be multiplied by this number. Note that it can also be a negative number."
```
Just remove `Note that it can also be a negative number` it is a scale factor, so no need for extended documentation. :)

zerothi commented on sisl/viz/plotly/plots/pdos.py:
```diff
@@ -39,7 +40,12 @@ class PdosPlot(Plot):
+        'species':          'atoms':          'orbitals':          'spin':
+        'normalize':          'color':          'linewidth':          'dash':
+        'split_on':          'scale': The final DOS will be multiplied by
+        this number. Note that it can also be a negative number. }
```
Also here, remove `Note that it can also be a negative number`

zerothi commented on sisl/viz/plotly/plots/pdos.py:
```diff
@@ -313,11 +326,14 @@ def _after_read(self):
-        self.spin_polarized = len(self.PDOS.shape) == 3
+        # Check if the PDOS contains spin resolution (there should be three dimensions,
+        # and the first one should be the spin components)
+        self.spin_polarized = self.PDOS.squeeze().ndim == 3
```
Are you also checking for non-colinear?
The last dimension may be either 1, 2 or 4?

pfebrer replied:
Do you have an example of a non-colinear PDOS file? What are each of the 4 spin channels in the non-colinear case?

zerothi replied:
See [here](https://github.com/zerothi/sisl/blob/293969f45ef21af2d63a04ca60a7dda5ab673008/sisl/io/siesta/pdos.py#L102)
and the following lines.

pfebrer replied:
Aah ok, is the sum channel just the sum of x, y and z? Or is it more complex?

zerothi replied:
No, it is more complex. The total DOS (sum) may be 4, while all the rest may be 0. You can think of it like a polarized calculation (here x and y are always 0). For a spin polarized up=1, down=1 one would have sum=2 and z = up - down = 0.
For up=1, down=0.5 one would have sum=1.5 and z=0.5.

So it is the *net* spin of that orbital. However one will always have `|x**2+y**2+z**2|**0.5 <= sum`.

pfebrer replied:
Ok, thanks! I have a non-colinear PDOS file and I'm trying to see what I need to change to make it work now :)

zerothi replied:
> By the way, I didn't use your `read_data(as_dataarray=True)` because it seemed like there wasn't an atom index, so I didn't understand. For two orbitals that have the same parameters but are in a different atom, is the DOS information summed?

No, true there isn't. But that could readily be added.
```python
a = sile.read_data(as_dataarray=True)
geom = a.attrs.geometry
atom_idx = geom.o2a(np.arange(geom.no))
```
done.
I would actually like to reduce number of coordinates since the stored matrix in `xarray` is a dense matrix. So adding a new coordinate needs careful thinking.
I am

pfebrer commented:
The scale can be used in any request! What I wrote there was just an example of how one may use it.

pfebrer commented:
Ok, let's see if you agree with the implementation:

Normally, if no spin was specified, I just aggregated the spin axis (either doing the mean or a sum, depending on `normalize`). However this doesn't make sense with non-colinear spins. Therefore, what I did is that when no spin is passed and the PDOS contains non-colinear information `"spin": ["sum"]` is added to the request so that it only takes that contribution by default.

So, if I grab a non-colinear PDOS file and split the PDOS on species:

```python
plot = sisl.get_sile("non-colinear_SrTiO3.PDOS").plot()
plot.split_DOS(on="species", name="$species")
```
I get the PDOS by species, using only the "sum" spin channel.

![image](https://user-images.githubusercontent.com/42074085/107504924-49133d80-6b9c-11eb-8fa1-841e2143a7f6.png)

You can see that the spin parameter of the request has been added:

![image](https://user-images.githubusercontent.com/42074085/107505310-ce96ed80-6b9c-11eb-94fc-9e84fa3a355b.png)

Also, of course if a value is passed for the spin parameter, that value is respected:

```python
plot.split_DOS(on="species", name="$species", spin=["x"])
```
![image](https://user-images.githubusercontent.com/42074085/107505554-2897b300-6b9d-11eb-9989-235ffb62c942.png)

Now, the only problem I saw with that implementation is that one desired to split the requests that they got further along spin, that would not be possible because there would always be the constraint of the spin channel being "sum":

![image](https://user-images.githubusercontent.com/42074085/107505960-b1aeea00-6b9d-11eb-8d70-55a2dee583fa.png)

So I added the argument `ignore_constraints` so that one can say: for the parameter I want to split on, please ignore the constraints of the existing query:

```python
plot.split_requests("Sr", on="spin", ignore_constraints=True)
```

![image](https://user-images.githubusercontent.com/42074085/107506151-fcc8fd00-6b9d-11eb-9b20-b04709a469a9.png)

One can choose to ignore also only specific constraints:

```python
plot.split_DOS(on="species", l=[0], name="$species, l=0")
plot.split_requests(0, on="spin+l", ignore_constraints=["spin"])
```

![image](https://user-images.githubusercontent.com/42074085/107506711-c93aa280-6b9e-11eb-8f7b-f14c9ab539ba.png)

You can see here the constraint about l being 0 is still respected, vs:

```python
plot.split_requests(0, on="spin+l", ignore_constraints=True)
```

![image](https://user-images.githubusercontent.com/42074085/107506871-06069980-6b9f-11eb-9885-24dba5933cd5.png)

Let me know if there's something that you don't like about the implementation and the way I tried to made spin splitting of the requests possible.

zerothi commented on sisl/viz/plotly/input_fields/dropdown.py:
```diff
@@ -155,9 +155,9 @@ class SpinSelect(DropdownInput):
         Spin.POLARIZED: [{"label": "↑", "value": 0}, {"label": "↓", "value": 1}],
-        Spin.NONCOLINEAR: [{"label": val, "value": val} for val in ("x", "y", "z")],
+        Spin.NONCOLINEAR: [{"label": val, "value": val} for val in ("sum", "x", "y", "z")],
         Spin.SPINORBIT: []
```
spin-orbit is the same as noncolinear

zerothi replied:
But what about here? why the `Spin.SPINORBIT` box? I guess this should have the same as noncolinear?

pfebrer replied:
Yes, I will add it, but for now it will have no effect on anything I believe.

zerothi commented on sisl/viz/plotly/input_fields/queries.py:
```diff
@@ -315,6 +337,13 @@ def _split_query(self, query, on, only=None, exclude=None, query_gen=None, **kwa
+
+            If `False`: all constraints considered.
+            If `True`: no constraints considered.
+            If array-like: parameters contained in the list ignore their constraints.
```
while this seems to solve the problem, I still don't get why ;)
I need to look into your request handles. Perhaps they could be restructured for the same effect with less clutter.

In particular it seems odd that the original one added l=1 when l=0 was requested explicitily.

pfebrer replied:
What do you mean the original one? `ignore_constraints=True` is not "the original one", the default value of `ignore_constraints` is `False`.

What seems odd about ignoring a constraint if the user told us to ignore it? hahaha

zerothi replied:
It just seems obvious that `l=0` would ignore different `l` values otherwise.

Again, you have to see me as a new user, and a new user would not see this connection :)

pfebrer replied:
Yes, that's why the **default behavior is to ignore the other `l` values**, therefore applying the constraint of `l=0` because the request that we are splitting imposes that constraint on `l`.

Only when you say `ignore_constraints=True` is it that `l` becomes free to be any value when you split on it.

I don't understand what the problem is :confused:

zerothi replied:
The problem is that users (including me) don't understand what `ignore_constraints` may be used for?

And to me it seems that ignore_constraints should always be `False`, when is `True` useful?

Also, from the doc's in `split_query` there is no other mention of `constraints`. So it is not clear what it refers to? Is it used together with `on` or something else? And also, how is it connected to `kwargs`.
:)

pfebrer replied:
> when is True useful

Mainly in non-colinear spin PDOS, that's why I introduced it. Automatically you get a request that has `spin=["sum"]` unless you specify a channel, because that's what you want (you don't want "sum" + "x" + "y", + "z"). But then if you split that on `spin`, you have the constraint that spin can only be `"sum"`, so basically you wouldn't split anything. That's why you need to ignore that constraint if you want the `x`, `y` and `z` channels for that request.

It may also be useful when you have a request like `{"atoms": [0, 1, 5, 8], "n": [3], "l": [0], "zeta": [1]}` and you want to see the DOS of that same request across `l`, for example. 

But the main point is to overcome the fact that splitting `spin=["sum"]` doesn't result in `spin=["x", "y", "z"]`.

> Also, from the doc's in split_query there is no other mention of constraints. So it is not clear what it refers to? Is it used together with on or something else? And also, how is it connected to kwargs.
> :)

In the `PdosPlot.split_requests` method:

> ignore_constraints: boolean or array-like, optional
>     
> determines whether constraints (**existing in the request to be splitted**) 
> on the parameters that we want to split along should be taken into consideration.
> If `False`: all constraints considered.
> If `True`: no constraints considered.
> If array-like: parameters contained in the list ignore their constraints.

How would you say it? Maybe "imposed by the request to be splitted"?

I don't know, this is a complex use case, and you will only come across it if you need to use it, where you will already understand what your problem is. Otherwise you don't even realize it's there.

zerothi replied:
It seems that I am missing the implementation details. I still have no idea what you are doing ;)

I'll have to read in the queries file regarding these details, I guess?

From the looks of it, it is very complicated :)
Why wouldn't users just add `spin=["sum", "x", ...]` explicitly when they want so?

pfebrer replied:
Yes I agree that the `OrbitalQueries` input is not simple, but that's the way I managed to handle all the corner cases and make queries behave like the user would expect. (note that is not used only by `PdosPlot`, also by `FatbandsPlot` and I'm quite happy with the consistency it achieved).

Maybe you find an easier way of handling all this :)

zerothi replied:
Ok. There is just lots to it :) And new for me, so lots of questions. It is very difficult to generalize and make things simple in a GUI interface like this :)
So all good :)

zerothi commented on sisl/viz/plotly/plots/pdos.py:
```diff
@@ -313,24 +326,33 @@ def _after_read(self):
+        if self.PDOS.squeeze().ndim == 3:
+            self.spin = {
+                2: sisl.Spin.POLARIZED,
+                4: sisl.Spin.NONCOLINEAR
```
and spin-orbit

pfebrer replied:
Soif I understand correctly, this implementation also makes sense for spin-orbit and there is no way to distinguish between spin-orbit and non-colinear from the shape of the PDOS array?

zerothi replied:
exactly :)

pfebrer replied:
Then I don't know what to do here, or if I should do anything :sweat_smile:

zerothi replied:
Are true. Here it is ok.

zerothi reviewed (changes requested):
For now this is fine. It seems to me the requests needs to figure this out automatically.
But lets start with this. :)

zerothi commented:
Is it complete now?

pfebrer commented:
I just have to add the tests for spin polarized and non-colinear PDOS files. Then it will be ready

pfebrer commented:
Done!

zerothi commented:
I'll merge in (with rebasing)
